beacon-chain/p2p/handshake.go

```diff
@@ -18,34 +18,36 @@ import (
 func (s *Service) AddConnectionHandler(reqFunc func(ctx context.Context, id peer.ID) error) {
 	s.host.Network().Notify(&network.NotifyBundle{
 		ConnectedF: func(net network.Network, conn network.Conn) {
-			if peerCount(s.host) > int(s.cfg.MaxPeers) {
+			log := log.WithField("peer", conn.RemotePeer().Pretty())
+			s.peers.Add(conn.RemotePeer(), conn.RemoteMultiaddr(), conn.Stat().Direction)
+			if len(s.peers.Active()) >= int(s.cfg.MaxPeers) {
+				log.Debug("We have enough peers; disconnecting")
 				if err := s.Disconnect(conn.RemotePeer()); err != nil {
 					log.WithError(err).Errorf("Unable to close peer %s", conn.RemotePeer())
-					return
 				}
+				return
 			}
 			multiAddr := fmt.Sprintf("%s/p2p/%s", conn.RemoteMultiaddr().String(), conn.RemotePeer().String())
 			log.WithFields(logrus.Fields{
-				"direction": conn.Stat().Direction,
-				"multiAddr": multiAddr,
-				"peerCount": peerCount(s.host),
-			}).Debug("Connection")
-			if peerstatus.IsBadPeer(conn.RemotePeer()) {
-				// Add Peer to gossipsub blacklist
+				"direction":   conn.Stat().Direction,
+				"multiAddr":   multiAddr,
+				"activePeers": len(s.peers.Active()),
+			}).Debug("Connection attempt")
+			if s.peers.IsBad(conn.RemotePeer()) {
+				// Add peer to gossipsub blacklist.
 				s.pubsub.BlacklistPeer(conn.RemotePeer())
-				log.WithField("peerID", conn.RemotePeer().Pretty()).Trace("Disconnecting with bad peer")
+				log.Trace("Disconnecting from bad peer")
 				if err := s.Disconnect(conn.RemotePeer()); err != nil {
```

Comment:
maybe add a connection state of `peers.PeerDisconnecting` before and `peers.PeerDisconnected` after.

Reply:
Same again :smile: we never attempted to connect at the protocol level so stay disconnected.